assets/training/finetune_acft_hf_nlp/environments/acpt/context/finetune_run.py

```diff
@@ -367,32 +369,117 @@ def add_task_specific_params(cmd: List[str], task_name: str, component_name: str
+                logger.info(f'in barrier file {barrier_file}')
+        finally:
+            portalocker.unlock(f)
+        with open(barrier_file, 'a+') as f:
+            while len(lines) < num_processes:
+                portalocker.lock(f, portalocker.LOCK_EX)
+                f.seek(0)
+                lines = f.readlines()
+                portalocker.unlock(f)
+                time.sleep(0.5)  # Polling interval
+        logger.info(f"Process {os.getpid()} has passed barrier with name {process_name}")
+
+
+def _run_subprocess_cmd(cmd: List[str], component_name: str, completion_files_folder: str,
+                        single_run=True, number_of_processes=1):
     """Run the subprocess command."""
     logger.info(f"Starting the command: {cmd}")
     completion_file = Path(completion_files_folder, f"{component_name}.complete.txt")
+    barrier_file = Path(completion_files_folder, f"{component_name}.barrier.txt")
+    Path(completion_files_folder).mkdir(parents=True, exist_ok=True)
     if completion_file.exists():
         logger.info(f"Skipping {component_name} as completion file exists: {completion_file}")
         return
-
-    # Not setting stdout and stderr will stream all the logs directly to stdout
-    process = subprocess.Popen(cmd)
-
-    # get the return code
-    return_code = process.wait()
-    if return_code != 0:
-        raise ACFTValidationException._with_error(
-            AzureMLError.create(
-                ACFTUserError,
-                pii_safe_message=(
-                    f"{component_name} failed"
+    process_name = os.environ.get('AZUREML_PROCESS_NAME', 'main')
+    if not barrier_file.exists():
+        Path(barrier_file).touch()
+        logger.info(f'Barrier file {barrier_file} is created by process name {process_name}')
+    if single_run:
+        if is_main_process():
+            logger.info(f"Executing command: {cmd} in single run mode. Process name is {process_name}")
+            # Not setting stdout and stderr will stream all the logs directly to stdout
+            process = subprocess.Popen(cmd)
+
+            # get the return code
+            return_code = process.wait()
+            if return_code != 0:
+                intermediate_folder = decode_output_from_env_var("intermediate_folder")
+                completion_files_folder = os.path.join(intermediate_folder, "completion_files")
+                shutil.rmtree(completion_files_folder, ignore_errors=True)
+                raise ACFTValidationException._with_error(
+                    AzureMLError.create(
+                        ACFTUserError,
+                        pii_safe_message=(
+                            f"{component_name} failed"
+                        )
+                    )
+                )
+            logger.info(f"{component_name} completed successfully")
+            Path(completion_files_folder).mkdir(parents=True, exist_ok=True)

```

Comment:
you already created this folder above so you can skip creating here.

Reply:
> you already created this folder above so you can skip creating here.

removing this , Thanks !